### PR TITLE
Mmdb server, closes #1779

### LINKED ISSUE #1779: [Analyzer] mmdb-server
## Name
mmdb-server

## Link
https://github.com/adulau/mmdb-server

## Type of analyzer
observable -> ip only

## Why should we use it
Free service available at https://ip.circl.lu/

## Possible implementation
 lookup via https://ip.circl.lu/geolookup/8.8.8.8

### PR DESCRIPTION
**Pull request by g4ze**
(Please add to the PR name the issue/s that this PR would close if merged by using a [Github]
Closes #1779

# Description

I have implemented the mmdb_server analyzer and have made the required migrations and documentation updates.
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [ ] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [x] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

RAW JSON:
{
    "id": 3,
    "user": {
        "username": "admin"
    },
    "tags": [],
    "comments": [],
    "pivots_to_execute": [],
    "permissions": {
        "kill": true,
        "delete": true,
        "plugin_actions": true
    },
    "analyzer_reports": [
        {
            "name": "Mmdb_server",
            "process_time": 1.5,
            "status": "SUCCESS",
            "end_time": "2024-01-18T20:30:27.422556Z",
            "parameters": {},
            "type": "analyzer",
            "id": 19,
            "report": [
                {
                    "ip": "191.121.10.0",
                    "meta": {
                        "build_db": "2023-11-20 12:50:37",
                        "nb_nodes": 1315572,
                        "db_source": "GeoOpen-Country",
                        "description": {
                            "en": "Geo Open MMDB database - https://github.com/adulau/mmdb-server"
                        }
                    },
                    "country": {
                        "iso_code": "BR"
                    },
                    "country_info": {
                        "Country": "Brazil",
                        "Alpha-2 code": "BR",
                        "Alpha-3 code": "BRA",
                        "Numeric code": "76",
                        "Latitude (average)": "-10",
                        "Longitude (average)": "-55"
                    }
                },
                {
                    "ip": "191.121.10.0",
                    "meta": {
                        "build_db": "2023-11-20 15:49:14",
                        "nb_nodes": 1316075,
                        "db_source": "GeoOpen-Country-ASN",
                        "description": {
                            "en": "Geo Open MMDB database - https://github.com/adulau/mmdb-server"
                        }
                    },
                    "country": {
                        "iso_code": "BR",
                        "AutonomousSystemNumber": "26615",
                        "AutonomousSystemOrganization": "TIM SA"
                    },
                    "country_info": {
                        "Country": "Brazil",
                        "Alpha-2 code": "BR",
                        "Alpha-3 code": "BRA",
                        "Numeric code": "76",
                        "Latitude (average)": "-10",
                        "Longitude (average)": "-55"
                    }
                }
            ],
            "errors": [],
            "start_time": "2024-01-18T20:30:25.918586Z"
        }
    ],
    "connector_reports": [],
    "pivot_reports": [],
    "visualizer_reports": [],
    "is_sample": false,
    "md5": "04bed81f0645696b4643ac953fd5da30",
    "observable_name": "191.121.10.0",
    "observable_classification": "ip",
    "file_name": "",
    "file_mimetype": "",
    "status": "reported_without_fails",
    "runtime_configuration": {
        "pivots": {},
        "analyzers": {},
        "connectors": {},
        "visualizers": {}
    },
    "received_request_time": "2024-01-18T20:30:25.588864Z",
    "finished_analysis_time": "2024-01-18T20:30:27.461155Z",
    "process_time": null,
    "tlp": "CLEAR",
    "errors": [],
    "warnings": [],
    "scan_mode": 1,
    "scan_check_time": null,
    "playbook_requested": null,
    "playbook_to_execute": null,
    "analyzers_requested": [
        "Mmdb_server"
    ],
    "connectors_requested": [],
    "analyzers_to_execute": [
        "Mmdb_server"
    ],
    "connectors_to_execute": [],
    "visualizers_to_execute": []
}

Screenshot:
image

### CODE REVIEW

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class MmdbServer(classes.ObservableAnalyzer):
+    base_url = "https://ip.circl.lu/geolookup/"
+
+    def run(self):
+        logger.info("nilay1")
+        observable_ip = self.observable_name
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/mmdb_server.py:19]: this assignment is useless

File: api_app/analyzers_manager/migrations/0057_analyzer_config_mmdb_server.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+from django.db import migrations
+from django.db.models.fields.related_descriptors import (
+    ForwardManyToOneDescriptor,
+    ForwardOneToOneDescriptor,
+    ManyToManyDescriptor,
+)
+
+plugin = {'name': 'Mmdb_server', 'python_module': {'module': 'mmdb_server.MmdbServer', 'base_path': 'api_app.analyzers_manager.observable_analyzers'}, 'description': 'mmdb-server is an open source fast API server to lookup IP addresses for their geographic location, AS number.', 'disabled': False, 'soft_time_limit': 60, 'routing_key': 'default', 'health_check_status': True, 'type': 'observable', 'docker_based': False, 'maximum_tlp': 'RED', 'observable_supported': ['ip'], 'supported_filetypes': [], 'run_hash': False, 'run_hash_type': '', 'not_supported_filetypes': [], 'health_check_task': None, 'model': 'analyzers_manager.AnalyzerConfig'}
```
Comment (mlodic) [inline api_app/analyzers_manager/migrations/0057_analyzer_config_mmdb_server.py:8]: please add a link to repo https://github.com/adulau/mmdb-server in the description

File: api_app/playbooks_manager/migrations/0023_add_mmdb_analyzer_free_to_use.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    pc = playbook_config.objects.get(name="FREE_TO_USE_ANALYZERS")
+    pc.analyzers.add(*["Mmdb_server"])
+    pc.full_clean()
+    pc.save()
+
+
+def reverse_migrate(apps, schema_editor):
+    playbook_config = apps.get_model("playbooks_manager", "PlaybookConfig")
+    pc = playbook_config.objects.get(name="FREE_TO_USE_ANALYZERS")
+    pc.analyzers.remove(*["Mmdb_serve"])
```
Comment (mlodic) [inline api_app/playbooks_manager/migrations/0023_add_mmdb_analyzer_free_to_use.py:19]: typo

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        logger.info(response)
+        return response.json()
+
+    @classmethod
+    def _monkeypatch(cls):
+        patches = [
+            if_mock_connections(
+                patch(
+                    "requests.get",
+                    return_value=MockUpResponse({}, 200),
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/mmdb_server.py:37]: please insert a sample in the mocked response here of a possible response of the mmdb server

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from api_app.analyzers_manager.exceptions import (  # AnalyzerConfigurationException
+    AnalyzerRunException,
+)
+from tests.mock_utils import MockUpResponse, if_mock_connections, patch
+
+logger = logging.getLogger(__name__)
+
+
+class MmdbServer(classes.ObservableAnalyzer):
+    base_url = "https://ip.circl.lu/geolookup/"
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/mmdb_server.py:15]: please configure this attribute to be a valid parameter (the user should be able to change it) https://intelowl.readthedocs.io/en/latest/Contribute.html#example-how-to-add-a-new-parameter-in-the-configuration-with-a-default-value

Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/mmdb_server.py:15]: noted

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    def run(self):
+        logger.info("nilay1")
```
[the remark this answers is not in the document]
Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/mmdb_server.py:18]: My bad; missed it.

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py (hunk 1 above)
Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/mmdb_server.py:19]: noted.

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py (hunk 4 above)
Comment (g4ze) [reply api_app/analyzers_manager/observable_analyzers/mmdb_server.py:37]: adding sample: https://github.com/adulau/mmdb-server?tab=readme-ov-file#lookup-of-an-ip-address

File: api_app/playbooks_manager/migrations/0023_add_mmdb_analyzer_free_to_use.py (hunk 3 above)
Comment (g4ze) [reply api_app/playbooks_manager/migrations/0023_add_mmdb_analyzer_free_to_use.py:19]: noted

File: api_app/analyzers_manager/migrations/0057_analyzer_config_mmdb_server.py (hunk 2 above)
Comment (g4ze) [reply api_app/analyzers_manager/migrations/0057_analyzer_config_mmdb_server.py:8]: noted!

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    base_url = str
+
+    observable_name: str
+
+    def run(self):
+        try:
+            response = requests.get(self.base_url + self.observable_name)
+            # response.raise_for_status()
+        except requests.RequestException as e:
+            logger.exception("Error while querying mmdb server")
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/mmdb_server.py:28]:
```suggestion
            logger.exception("Error while querying mmdb server: {e}")
```

File: api_app/analyzers_manager/observable_analyzers/mmdb_server.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    base_url = str
+
+    observable_name: str
+
+    def run(self):
+        try:
+            response = requests.get(self.base_url + self.observable_name)
+            # response.raise_for_status()
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/mmdb_server.py:26]:
```suggestion
            response.raise_for_status()
```

**Review by mlodic — CHANGES_REQUESTED** [review]
almost done, it is remained to check the CI checks and solve them

**Comment by g4ze** [comment]
Hello @mlodic ! Hope you're doing well. I think, I have made the required corrections for consistent code quality.

**Comment by mlodic** [comment]
please pull from develop and adjust the migration codes. This happens when other migrations are merged in the `develop` repo while you are working. Then it's done.

**Comment by g4ze** [comment]
hello there,
(minor update)
I'm currently facing some issue regarding build/running intelOwl; i believe it's a network problem. Although my internet connection speed at my university is pretty high, I can't seem to find out how to get rid of such TLS handshake and timeout errors. It has worked pretty seamlessly before on the same network. 
'''
~/Documents/GitHub/intel-owl dev 2/IntelOwl-dev$ ./start test up -- --build
[+] Running 2/2
 ✘ celery_worker_default Error                                    9.6s 
 ✘ celery_beat Error                                             17.9s 
[+] Building 42.3s (4/4) FINISHED                       docker:default
 => [uwsgi internal] load build definition from Dockerfile        0.0s
 => => transferring dockerfile: 3.31kB                            0.0s
 => [uwsgi internal] load metadata for docker.io/library/python:  0.0s
 => ERROR [uwsgi internal] load metadata for docker.io/library/  42.2s
 => [uwsgi auth] library/node:pull token for registry-1.docker.i  0.0s
------
 > [uwsgi internal] load metadata for docker.io/library/node:lts-alpine3.18:
------
failed to solve: node:lts-alpine3.18: failed to do request: Head "https://registry-1.docker.io/v2/library/node/manifests/lts-alpine3.18": net/http: TLS handshake timeout
'''
This issue might take some more time to resolve.

**Comment by drosetti** [comment]
Hi @g4ze, this error is my fault: I fixed a bug in develop and I didn't respect the linting rules. Now I fixed it, pull again from develop and let me know if you have other problems! :)

**Comment by mlodic** [comment]
one last change, please pull from `develop` anche change your migration number to 61 (and its dependency) because another similar PR has been merged recently. Then I think we are done

**Comment by g4ze** [comment]
done.

**Comment by mlodic** [comment]
from the tests, it seems that there is another problem in the migration ,related to recent changes introduced by @0ssigeno. @g4ze you are done, we'll fix this and merge the PR. Ty!

**Comment by mlodic** [comment]
> Can you guide me to any other issue I can take up?

You could try something a little different which is to rework an old analyzer:
https://github.com/intelowlproject/IntelOwl/issues/439

**Comment by 0ssigeno** [comment]
Yeah, you made the migration before my fix. I will make the changes.

**Comment by 0ssigeno** [comment]
Tbh the error is how you wrote the adding of the mmdb server to the playbook.
No issues, I will fix it.

**Comment by g4ze** [comment]
thankyou for the fix! I will keep the changes in mind now onwards.

**Comment by 0ssigeno** [comment]
I have added a line in the docs to help future contributors to correctly set up the migration file for the playbook.

**Review by 0ssigeno — APPROVED** [review]
lgtme

**Pull request merged**